internal/aws/k8s/k8sclient/endpoint.go

```diff
@@ -0,0 +1,205 @@
+// Copyright  OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8sclient
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"go.uber.org/zap"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/k8s/k8sutil"
+)
+
+const (
+	TypePod = "Pod"
+)
+
+type Service struct {
+	ServiceName string
+	Namespace   string
+}
+
+func NewService(name, namespace string) Service {
+	return Service{ServiceName: name, Namespace: namespace}
+}
+
+type EpClient interface {
+	// Get the mapping between pod key and the corresponding service names
+	PodKeyToServiceNames() map[string][]string
+	// Get the mapping between the service and the number of belonging pods
+	ServiceToPodNum() map[Service]int
+
+	// shutdown is only used internally by clientset to stop the EpClient
+	shutdown()
+}
+
+type epClientOption func(*epClient)
+
+func epSyncCheckerOption(checker initialSyncChecker) epClientOption {
+	return func(e *epClient) {
+		e.syncChecker = checker
+	}
+}
+
+type epClient struct {
+	sync.RWMutex
```

Comment:
Is this pattern of directly inheriting from `RWMutex` common? I don't know if I've seen it before, in Go there shouldn't be overhead of making it a field.